File: Java/src/Remove Duplicates from Sorted List II .java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
public class Solution
{
  public ListNode deleteDuplicates(ListNode head)
  {
    if (head == null)
      return head;
    ListNode dummy = new ListNode(-1);
    dummy.next = head;
    ListNode cur = dummy;
    ListNode ret = dummy;
    boolean isDup = false;

    if (head.val == dummy.val)
      isDup = true;

    while (cur.next.next != null)
    {
      if (cur.val != cur.next.val && cur.next.val != cur.next.next.val)
      {
        ret.next = cur.next;
        ret = ret.next;
      }
      cur = cur.next;
    }

    if (cur.next != null && cur.val != cur.next.val)
    {
      ret.next = cur.next;
      ret = ret.next;
    }
    else if (cur.next != null && cur.val == cur.next.val)
    {
      ret.next = cur.next.next;
      ret = ret.next;
    }

    if (isDup)
    {
      return dummy;
    }
    else
    {
      return dummy.next;
    }
  }
}

File: Python/Remove Duplicates from Sorted List II .py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        if not head:
            return None
        dummy = ListNode(-1)
        dummy.next = head
        cur = dummy
        ret = dummy
        isDup = False
        
        if head.val == dummy.val:
            isDup = True

        while cur and cur.next and cur.next.next:
            if cur.val != cur.next.val and cur.next.val != cur.next.next.val:
                ret.next = cur.next
                ret = ret.next
            cur = cur.next
        
        if cur.next and cur.val != cur.next.val:
            ret.next = cur.next
            ret = ret.next
        elif cur.next and cur.val == cur.next.val:
            ret.next = cur.next.next
            ret = ret.next
        if isDup:
            return dummy
        else:
            return dummy.next
                
                
                
        